### ISSUE TITLE
The default dtype for creating datasets does not match the default dtype for creating numpy arrays

### ISSUE THREAD

**Issue by siebenkopf** [h5py/h5py#911]
I recently stepped into an issue, which was unexpected. I used `h5py` to write a `numpy` array, which I had created with its default datatype, which is `numpy.float64'. After writing and loading with `h5py`, the datatype unexpectedly changed to `numpy.float32`. A code snippet to replicate this issue is:

```
import numpy, h5py
a = numpy.random.randn(10)
h = h5py.File("test.hdf5")
d = h.create_dataset("X", (10,))
d[:] = a
b = d[:]
print (a.dtype, b.dtype)
print numpy.allclose(a,b,atol=1e-12,rtol=1e-12)
```

This snippet will print `float64 float32` and `False`.
As you can see, I am using default datatypes both for `numpy` and `h5py`, which differ.

There are two possible solutions to solve this issue:
1. Change the default datatype of `File.create_dataset` to `numpy.float64`
2. Add a documentation that the default of `File.create_dataset` is `numpy.float32`. At the moment, the documentation just says: "defaults to `'f'`": http://docs.h5py.org/en/latest/high/dataset.html#creating-datasets while it is not clear that `'f'` means `numpy.float32`.

I guess that 2. is the easier solution. However, it should be made clear, maybe by adding a warning, that this is not the default for `numpy.ndarray`'s.

**Comment by aragilar** [comment]
1 is not a solution, that breaks backward compatibility badly.

`'f'` comes from numpy (`np.dtype('f')` maps to `numpy.float32`), as does `'i8'` used in the next example (dtypes use numpy types, unless it's a special dtype (which still use numpy dtypes interally)).

A PR which makes this point more explicit (that dtypes arguments are interpreted by numpy) and emphases it with an example would be good, but a runtime warning would not be useful in this case I feel.

**Comment by siebenkopf** [comment]
I agree that 1. was just an idea, but would not be applicable.

From the documentation, I did understand that the data type was interpreted by `numpy`. However, without reading the `numpy` documentation, it was not clear that `'f'` maps to `numpy.float32`, instead of the `numpy` default `numpy.float64`.

I did not want to add a runtime warning, but a `.. warning::` inside the documentation.

I am not sure, though, if I will find the time for a PR.

**Comment by tacaswell** [comment]
If you pass the data into `create_dataset`

```python
In [10]: g = f.create_dataset('b', (10,), data=np.arange(10, dtype=float))

In [11]: g.dtype
Out[11]: dtype('float64')
```

or create the dataset implicitly

```python
In [12]: f['c'] = np.linspace(0, 5, 10)

In [13]: f['c'].dtype
Out[13]: dtype('float64')
```

It will correctly detect the type

I think the only thing needed here is for the `create_dataset` docs to be clearer that it defaults to float32.

**Comment by siebenkopf** [comment]
Agreed. Thanks for the additional info on how to easily create datasets. I think, these should be mentioned on the first page (the quick start guide), too. These are much easier to use (you don't need to provide the `shape` or the `dtype` when using the implicit dataset generation) and are definitely more pythonic.

**Comment by NumesSanguis** [comment]
If you're clarifying on data types anyway, could you make a page with which values are accepted for `dtype`? The only page having an overview of data types is: [http://docs.h5py.org/en/latest/faq.html#what-datatypes-are-supported](http://docs.h5py.org/en/latest/faq.html#what-datatypes-are-supported)

Nowhere is mentioned how you specify what argument to pass. I only figured out that `'f4'` can be used for 32-bit float because I saw in the docs `'i8'` and guessed you had to use the first letter.

Also, cross-referencing of `dtype` is a mess:

[http://docs.h5py.org/en/latest/high/dataset.html?highlight=dtype#creating-datasets](http://docs.h5py.org/en/latest/high/dataset.html?highlight=dtype#creating-datasets) mentions that you have to specify `dtype`, but doesn't give a link where to find the types.

and probably on the [Quick Start Guide](http://docs.h5py.org/en/latest/quick.html?highlight=dtype) it would also be nice to have a cross-link to an overview of `dtype`s.

**Comment by siebenkopf** [comment]
I am not sure how #959 closes this issue. There is no mentioning of default datatypes of `numpy` and `h5py` in #959.
